gpxpy/parser.py

```diff
@@ -165,7 +165,7 @@ def __init__(self, xml_or_file=None, parser=None):
 
     def init(self, xml_or_file):
         text = xml_or_file.read() if hasattr(xml_or_file, 'read') else xml_or_file
-        if text[:3] == "\xEF\xBB\xBF": #Remove utf-8 Byte Order Mark (BOM) if present
+        if text.startswith(u'\xEF\xBB\xBF'):  # Remove utf-8 Byte Order Mark (BOM) if present
```

Comment:
> From what I see, it can either accept a file or a string-like object. A file.read() is bytes in Python 2 and unicode in Python 3. But when it's a string-like object it can be anything in both Py 2 and 3.

For `file.read()`, it depends on how the file was opened.

> What do you think should be the right choice here? Is lxml handling bom encoding fine?

I did a quick test, and I think `lxml` does work correctly. What I did was
```
$ cp route.gpx route-bom.gpx
$ vim route-bom.gpx
:set bomb
:wq

$ ipython3
> from lxml import etree
> with open('route.gpx', 'rb') as f:
>     XML(f.read())
> with open('route.gpx', 'rb') as f:
>     dom = etree.XML(f.read())
> with open('route-bom.gpx', 'rb') as f:
>     dom_bom = etree.XML(f.read())
```
and compared them using [this function](http://stackoverflow.com/a/24349916).

> For lxml it seems that it is converting back to bytes in line 101.

Hmm, you're right. I think this stuff is mostly compatibility for the `minixml` implementation, which unfortunately takes a `str` instead of `bytes`. But doing a double decode->encode is not the best choice.

I think firstly, this function should be modified to pick one or the other (preferably `bytes`) and secondly, all the special decoding can be moved into the `minixml` implementation.